### ISSUE TITLE
Support optional inputs and outputs specified via @Input/@Output + @Optional decorators

### ISSUE THREAD

**Issue by JamesHenry** [ngParty/ng-metadata#45]
First off - brilliant project, thanks for all the hard work! I am already successfully using this on a large Angular 1.x app.

Is there any particular reason why optional inputs and outputs are not supported via the decorator style? I.e. The DDO syntax of `[..scope/bindToController..]: { myOptionalInput: '=?', myOptionalOuput: '&?'  }`

If not it would make a great addition!

Many thanks.

**Comment by Hotell** [comment]
Hey, thanks for kind words! Yes I know that optional is missing. 

I wanted to use `@Optional` for this but that could be miss leading in comparison with angular 2, which is using Optional solely for optional DI. 

There are 3 ways how we can solve this:
- use `@Optional` for both Directive DI in constructor and for scope bindings with `@Input`,`@Output`,`@Attr` 
- create some special decorator just for bindings (`@Opt` for example?)
- make all bindings `?` optional by default

IMO I would go with that last one, wdyt @JamesHenry ?

Anyway you can apply optional bindings today via alias (it's cumborsome but works hehe)

```typescript
@Component({...})
class MyComponent{
  @Input('?myData') myData;
  @Output('?onDataChange') onDataChange: Function;
}
```

```html
<my-cmp my-data="$ctrl.data" on-data-change="$ctrl.changeMe(newData)"></my-cmp>
```


under the hood it will create following bindToController map:
```typescript
{
  myData: '=?myData',
  onDataChange: '&?onDataChange'
}
```

**Comment by JamesHenry** [comment]
Ah, that makes a lot of sense then.

Thanks for the workaround, it is more succinct than what I had done, which is mix the use the decorators for non-optional bindings, and the `inputs: []` array for optional ones! Either way works, but I prefer the alias one for now.

Making them all optional would definitely make this situation easier, but thinking about it from the other angle (of wanting mandatory bindings)...

Even though the Angular error messaging around it is not very clear, having bindings as non-optional can be useful at development time for catching improperly configured components (for those "oh I forgot to specify "my-prop=" moments :) ). That could catch some developers out.

Tbh I think the aliasing solution is probably the best:
- It is explicit
- It maps best to what is expected in ng2

Simply adding an example of this case to the docs would be enough IMO

**Comment by Hotell** [comment]
Yeah I'll leave it as is, just will add that *how to* to docs.

Anyway, I think you are not doing what you think you are doing, with optional bindings :)

**tl;dr:**
optional binding `?` will not throw error or anything if you don't provide it on your component/directive html ( the mechanism are not working like react propTypes for example )

If you look at angular docs, you'll find following:
https://docs.angularjs.org/api/ng/service/$compile#-scope-
> Optional attributes should be marked as such with a question mark: =? or =?attr. If the binding expression is non-assignable, or if the attribute isn't optional and doesn't exist, an exception ($compile:nonassign) will be thrown upon discovering changes to the local value, since it will be impossible to sync them back to the parent scope

I made a quick bin to demonstrate: 
https://jsbin.com/zufoqok/2/edit?html,js,output

After you agreement I will close this issue.

Lemme know wdyt,
thanks!

**Comment by JamesHenry** [comment]
Just a misunderstanding I think - you have just proven what I was trying to call your attention to :D

You suggested:
> make all bindings ? optional by default

...and as you say above:
> optional binding ? will not throw error or anything if you don't provide it on your component/directive 

...and what I said is that:
> having bindings as **non**-optional can be useful at development time for catching improperly configured components

For example, in your `<consumer>`, I am saying that it is *a good thing* that the bindings are not optional because if the developer misses out the bindings he has specified as being required, Angular will throw and the developers gets that feedback quicker than a bug report further down the road :)

So I was saying that, if you were to solve this issue by "make all bindings ? optional by default", you would be silently suppressing this behaviour and developers would not get the same feedback from errors if the bindings were not specified.

Anyway, we definitely both agree on the resolution for the issue, so feel free to close.

Thanks again for the great lib!

**Comment by Hotell** [comment]
Haha yeah, sorry my bad :)

I'm glad you like it! 1.0 is on it's way!

thanks :)
